nixos/modules/services/misc/rmfakecloud.nix

```
@@ -0,0 +1,119 @@
+
+      storageUrl = mkOption {
+        type = types.str;
+        example = "https://local.appspot.com";
+        description = ''
+          Controls whether file upload/download goes through the local proxy
+          or to an external server. It's the address of rmfakecloud as visible
+          from the tablet, especially if the host is behind a reverse proxy or
+          in a container.
+        '';
+      };
+
+      port = mkOption {
+        type = types.port;
+        default = 3000;
+        description = ''
+          Listening port number.
+        '';
+      };
+
+      logLevel = mkOption {
+        type = types.enum [ "info" "debug" "warn" "error" ];
+        default = "info";
+        description = ''
+          Set the log verbosity.
+        '';
+      };
+
+      extraSettings = mkOption {
+        type = with types; attrsOf str;
+        default = { };
+        example = { RM_HTTPS_COOKIE = true; };
+        description = ''
+          Extra settings in the form of a set of key-value pairs.
+          For tokens and secrets, use `environmentFile` instead.
+        '';
+      };
+
+      environmentFile = mkOption {
+        type = with types; nullOr path;
+        default = null;
+        example = "/etc/secrets/rmfakecloud.env";
+        description = ''
+          Path to an environment file.
+          This can be used to securely store tokens and secrets outside of the
+          world-readable Nix store.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.rmfakecloud = {
+      description = "rmfakecloud remarkable self-hosted cloud";
+
+      environment = {
+        STORAGE_URL = cfg.storageUrl;
+        PORT = toString cfg.port;
+        LOGLEVEL = cfg.logLevel;
+      } // cfg.extraSettings;
+
+      preStart = ''
+        # Generate the secret key used to sign client session tokens.
+        # Replacing it invalidates the previously established sessions.
+        if [ ! -f jwt_secret_key ]; then
+          (umask 077; touch jwt_secret_key)
+          ${pkgs.openssl}/bin/openssl rand -base64 48 >> jwt_secret_key
+        fi
+      '';
+
+      script = ''
+        if [[ ! -v JWT_SECRET_KEY ]]; then
+          export JWT_SECRET_KEY=$(cat jwt_secret_key)
+        fi
+
+        ${pkgs.rmfakecloud}/bin/rmfakecloud
+      '';
+
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "network-online.target" ];
+      after = [ "network-online.target" ];
+
+      serviceConfig = {
+        Type = "simple";
+        Restart = "always";
+
+        EnvironmentFile =
+          mkIf (cfg.environmentFile != null) cfg.environmentFile;
+
+        AmbientCapabilities =
+          mkIf (cfg.port < 1024) [ "CAP_NET_BIND_SERVICE" ];
+
+        DynamicUser = true;
+        PrivateDevices = true;
+        ProtectHome = true;
+        ProtectKernelTunables = true;
+        ProtectKernelModules = true;
+        ProtectControlGroups = true;
```

The following is solely opinion-based of what I'd like to see here. AFAIK, nixpkgs has no defined threat model modules should align themselves to nor a policy on recommended security features modules should use.

```suggestion
        ProtectControlGroups = true;
        CapabilityBoundingSet = [ "" ];
        DevicePolicy = "closed";
        LockPersonality = true;
        MemoryDenyWriteExecute = true;
        ProtectClock = true;
        ProtectHostname = true;
        ProtectKernelLogs = true;
        ProtectProc = "invisible";
        ProcSubset = "pid";
        RemoveIPC = true;
        RestrictAddressFamilies = [
          "AF_INET"
          "AF_INET6"
        ];
        RestrictNamespaces = true;
        RestrictRealtime = true;
        RestrictSUIDSGID = true;
        SystemCallArchitectures = "native";
```

---

For a network-facing service I'd go as far as suggest using `systemd.services.rmfakecloud.confinement` which works well for this service: http://cgit.notk.org/asmadeus/nixos-config.git/tree/profiles/rmfakecloud.nix
(But happy to keep it locally though, it doesn't like DynamicUser...)